src/main/scala/com/github/jberggg/router/Service.scala

```diff
@@ -0,0 +1,42 @@
+package com.github.jberggg.router
+
+import cats.implicits._
+import org.http4s.Uri.Path
+
+import org.scalajs.dom.window
+import cats.effect.std.Dispatcher
+import org.scalajs.dom.Event
+import fs2.concurrent.Channel
+import fs2._
+import cats.effect.kernel.{Async, Resource}
+
+object Service {
+  
+    def setupInfrastructure[F[_] : Async ]: F[Channel[F, Path]] = for {
+        channel     <- Channel.unbounded[F,Path]
+        location    <- getCurrentLocation[F]
+        _           <- channel.send(location)
+    } yield channel
+
+    private[router] def registerEventHandlerAndToStream[ F[_] : Async ](c: Channel[F, Path]): Stream[F, Path] = for {
+        dispatcher <- Stream.resource(Dispatcher[F])
+        callback   =  (_: Event) => dispatcher.unsafeRunAndForget( getCurrentLocation[F].flatMap( l => c.send(l) ) )
```

Comment:
Looking at the docs for `hashchange` it looks like the `HashChangeEvent` actually carries information about the new/old URLs. So it seems like you can directly read it off of the event instead of calling `getCurrentLocation`.

https://developer.mozilla.org/en-US/docs/Web/API/Window/hashchange_event

https://github.com/scala-js/scala-js-dom/blob/3600c24f0612c940e59a5ff7371491e9eb0f9100/dom/src/main/scala/org/scalajs/dom/HashChangeEvent.scala

Reply:
I changed the callback to use the supplied path in the event. Because it is the full path I need to calculate the hash part though but I guess it is still more efficient then to call a browser API.